scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala

```diff
@@ -63,9 +65,9 @@ class SparkWriter(val sparkMode: SparkMode) extends Writer {
           val newForced = forcedPipes + ((c, opt) -> workVal)
           val newInitToOpt = initToOpt + ((c, init) -> opt)
 
-          (copy(sources = newSources, forcedPipes = newForced, initToOpt = newInitToOpt), true)
+          (copy(sources = newSources, forcedPipes = newForced, initToOpt = newInitToOpt), true, forcedRdd)
         case Some(_) =>
```

Comment:
I wonder if your failing test is due to this comment? You aren't returning a previously forced value.

Reply:
I have not looked at that yet. But one other thing I suspect: the counter increment is done through map transform on RDD. Spark does not guarantee consistency in transformation (only actions).

is we want strict consistency we would need a blocking action like `forEach`

Reply:
the reason I suspect this is because the result is a strict multiple of the expected value (200 returned vs 100 expected). This leads me to think that the map phase was recomputed twice by the spark engine.

Reply:
Yes, but it could also be because the planner asked Spark to compute it twice. Not clear to me. I would investigate and try to falsify both.

Reply:
> shouldn't we get the `forcedRdd` from the `WorkVal` you have here? I think so. If not, please add a comment why not.

I am a bit confused as to what is being referred to here. `WorkVal` is just `val workVal: WorkVal[T] = (src, forcedRdd)` no?.

Anyways it does appear that we have a double computation somewhere in the planner. Specifically happens in the case that this line is included: `.flatMap(_ => TypedPipe.from(sinkExample).toIterableExecution)` after an execution.

Reply:
yeah `WorkVal` is that pair, but if you have a cached value, you are ignoring the cached value and returning the input. I think you need to use that cached value. Don't you? If not, can you explain why we can safely ignore an existing `Future[RDD[_ <: A]]` given that we already have another equivalent computation?

I don't see why, but it could be true.

Basically: if we have two identical optimized typedpipes, they should share the exact same `Future[RDD...`.

Reply:
ok yeah if I'm understanding correctly we should do this. Sounds good to me if this is the case: 

```scala
-        case Some(_) =>
-          (copy(initToOpt = initToOpt + ((c, init) -> opt)), false, rdd)
+        case Some(cacheHit) =>
+          (copy(initToOpt = initToOpt + ((c, init) -> opt)), false, cacheHit._2)

```

Reply:
yeah, I think so, or I would write `case Some((_, frdd)) => ...`